### PR TITLE
[ENH] Refactor of classifier base class: input checking and conversion

### PR DESCRIPTION
**Pull request by TonyBagnall**
#### Reference Issues/PRs

https://github.com/alan-turing-institute/sktime/issues/1537

#### What does this implement/fix? Explain your changes.
this addresses some problems with check_X_y/check_X and integrates the capabilities tags into the new fit/predict structture. It does this in four stages
1. check the input is a valid data type: function check_classifier_input (utils.validation.panel)  
2. gets the data characteristics: missing values, multivariate and unequal length. function get_data_characteristics  (utils.validation.panel)  
3. checks the classifier can handle the data characteristics. member function check_capabilities (classification.base)
4. performs any conversions from DataFrame to array or vice versa member function convert_input (classification.base)

#### Does your contribution introduce a new dependency? If yes, which one?
no

#### What should a reviewer concentrate their feedback on?

whether to use datatypes conversion or not, whether the y conversions are ok, testing

### CODE REVIEW

File: sktime/classification/__init__.py
Comment (MatthewMiddlehurst) [inline sktime/classification/__init__.py:2]: This is for the whole classification package, right? Not just tests.

A bit of a nitpick.

File: sktime/classification/base.py
Comment (MatthewMiddlehurst) [inline sktime/classification/base.py:66]: May need to add this to all the classifier docs? Maybe not now but a potential issue

File: sktime/classification/base.py
Comment (MatthewMiddlehurst) [inline sktime/classification/base.py:278]: typo, same one in some other methods also

File: sktime/classification/kernel_based/_arsenal.py
Comment (MatthewMiddlehurst) [inline sktime/classification/kernel_based/_arsenal.py:182]: Why is this in Arsenal and no other class?

File: sktime/registry/_tags.py
Comment (MatthewMiddlehurst) [inline sktime/registry/_tags.py:233]: Doesn't this and the numpy version still exist for other modules? Don't think they should be removed, if so.

**Review by MatthewMiddlehurst — COMMENTED** [review]
Good stuff. Couple of comments:

The convert methods are currently inherited by all classifier, they are private anyway I know, but should that be the case? Could just be a function in the file, possibly.

convert_y_to_numpy is the default, but all the changes to class tags set it again. The classes that use dataframes do not convert to series.

File: sktime/classification/__init__.py
Comment (TonyBagnall) [reply sktime/classification/__init__.py:2]: Good spot, tganks

File: sktime/classification/base.py
Comment (TonyBagnall) [reply sktime/classification/base.py:66]: Indeed, will do so

File: sktime/classification/kernel_based/_arsenal.py
Comment (TonyBagnall) [reply sktime/classification/kernel_based/_arsenal.py:182]: Legacy of debugging :)

File: sktime/registry/_tags.py
Comment (TonyBagnall) [reply sktime/registry/_tags.py:233]: I'm pretty sure it has been removed from other modules, and tags test should verify this, but I'll double check

**Comment by TonyBagnall** [comment]
Re the conversion, I think Franz wants to convert it to data types, so I've just put them here, it was a good learning exercise for me. Convert needs access to tags which use the shadowed _tags, so need access to the calling object. Could pass the classifier explicitly in a static method I guess? Ideally these would be final

**Comment by TonyBagnall** [comment]
in fact it needs a doc string for attributes. This is not going in release 9 so no rush here. @MatthewMiddlehurst regarding _threads_to_use, is this not called n_jobs elsewhere? Also, should _class_dictionary not be class_dictionary_, or _class_dictionary_?!

File: sktime/registry/_tags.py
Comment (MatthewMiddlehurst) [reply sktime/registry/_tags.py:233]: Is the registry actually used elsewhere? I thought it was just to inform people about the types of tags.

**Comment by MatthewMiddlehurst** [comment]
Yes, all the input for threading is n_jobs, but the actual amount being used can be different from that (i.e. -1 == all threads). The other naming option was just a private version, _n_jobs, but I thought that could be confusing. Don't mind if you change any of the names.

File: sktime/registry/_tags.py
Comment (TonyBagnall) [reply sktime/registry/_tags.py:233]: I think its part of the tests to check estimators have valid tags, in registry/tests/test_lookup

**Comment by TonyBagnall** [comment]
> Yes, all the input for threading is n_jobs, but the actual amount being used can be different from that (i.e. -1 == all threads). The other naming option was just a private version, _n_jobs, but I thought that could be confusing. Don't mind if you change any of the names.

@MatthewMiddlehurst maybe we should pass and set n_jobs in the base class constructor, defaulting to 1? Looks a bit odd testing its existence in base where its not defined?

**Comment by TonyBagnall** [comment]
in addition to n_jobs, the random_state and vebose could move up to base. Not sure really, maybe get this version in then discuss at a developer meeting

**Comment by TonyBagnall** [comment]
I've checked clustering, no use of deleted tags (anda similar restructuring to this class in progress by @chrisholder). In classification, it is not used by the  _ensemble or the _column_ensemble retro classifiers, nor KKN/PF.  Annotation Alignment and Regression dont use tags yet. Transformations does use tags but not this one. So I think its all clear to remove it.

**Comment by MatthewMiddlehurst** [comment]
coerce-X-to-numpy is the more concerning one really, used by  ["forecaster", "classifier", "regressor"]

**Comment by TonyBagnall** [comment]
> coerce-X-to-numpy is the more concerning one really, used by ["forecaster", "classifier", "regressor"]

that is the one I meant :) Its actually not used any more, despite that listing, unless you can see it somewhere?

File: sktime/classification/base.py
Comment (chrisholder) [inline sktime/classification/base.py:50]: These are parameters not attributes?

File: sktime/classification/base.py
Comment (chrisholder) [inline sktime/classification/base.py:289]: These are parameters not attributes?

File: sktime/classification/base.py
Comment (chrisholder) [inline sktime/classification/base.py:391]: This should be parameters not arguments

File: sktime/classification/base.py
Comment (chrisholder) [inline sktime/classification/base.py:551]: This whole function can be done using .any

```
def _has_nans(x: np.ndarray) -> bool
    return np.isnan(x).any()
```

File: sktime/classification/interval_based/_risf.py
Comment (chrisholder) [inline sktime/classification/interval_based/_risf.py:32]: You don't need to keep this if it has been moved here?

File: sktime/utils/data_io.py
Comment (chrisholder) [inline sktime/utils/data_io.py:67]: Why are you undeprecating rather than using the on it suggests (as of 0.10) should be in datasets/_data_io

File: sktime/utils/validation/panel.py
Comment (chrisholder) [inline sktime/utils/validation/panel.py:14]: Remove

**Review by chrisholder — CHANGES_REQUESTED** [review]

File: sktime/classification/base.py
Comment (TonyBagnall) [reply sktime/classification/base.py:551]: I'm sure there was a reason I didnt want to do that, possibly for the early abandon? I imagine franz will replace this anyway, so I will leave it for now but not commit before sorting this out (I'll run some tests if necessary)

File: sktime/classification/interval_based/_risf.py
Comment (TonyBagnall) [reply sktime/classification/interval_based/_risf.py:32]: it does not conform to the new interface, so unless someone want to do that, I will delete it. I'll contact the code owner

File: sktime/utils/data_io.py
Comment (TonyBagnall) [reply sktime/utils/data_io.py:67]: thanks chris, I deleted the wrong deprecation warning!

Comment (TonyBagnall) [reply sktime/utils/data_io.py:67]: this is odd, I'm not sure what happened to the warnings, I'll put them back in

**Comment by TonyBagnall** [comment]
@MatthewMiddlehurst "convert_y_to_numpy is the default, but all the changes to class tags set it again. The classes that use dataframes do not convert to series." these are the ones that seem to reset it. I'm not sure which classifiers actually expect Series, I think it is the old unconverted ones (or MUSE?). All test pass, probably doing their own internal conversion, but need to resolve this. If anyone knows which need Series for y, let me know and save me a trawl through the codebase!

**Comment by MatthewMiddlehurst** [comment]
I dont think anything requires series a Series y going by the tests, I would set the DataFrame classifiers (MUSE and the 3 PF classes) to use them by default though. Or not, works currently.

**Comment by fkiraly** [comment]
FYI, see #1749 - I've refactored this to use `datatypes`.
Is that much simpler or not :-)

**Comment by fkiraly** [comment]
PS: the testing logic for the data characteristics is now in `datatypes`, the tests that fail since the functions no longer exist in the base classifier should be removed.

File: sktime/classification/base.py
Comment (TonyBagnall) [reply sktime/classification/base.py:551]: so I remember now, the problem with np.isnan is it returns an array of booleans, which a) is wasteful in memory and b) does not allow the early abandon with the first nan. But I am not really fussed either way, its unlikely to be a huge overhead

Comment (fkiraly) [reply sktime/classification/base.py:551]: I've replaced this by faster functionality in the PR I made, have you looked at that?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Just for complying with the formal rules:
as per discussion on slack and asked by @TonyBagnall, I've made a PR to this PR which replaces checking/conversion functionality in this PR with already existing checking/conversion functionality in the `datatypes` module, see here:
https://github.com/alan-turing-institute/sktime/pull/1749

All checking/conversion except the estimator capability checking already exists on `main`, so merging this PR as is would lead to substantial duplication of checking/conversion functionality. We need to avoid that and resolve this before this gets merged.

I'm not picky on structure, tag names, etc. The main blocker for me would be *not duplicating the checking/conversion logic*.

File: sktime/classification/base.py
Comment (fkiraly) [inline sktime/classification/base.py:150]: why is this incorrect?

File: sktime/classification/feature_based/_signature_classifier.py
Comment (fkiraly) [inline sktime/classification/feature_based/_signature_classifier.py:105]: this is not a tag that exists, is it?

File: sktime/classification/interval_based/_risf.py
Comment (fkiraly) [inline sktime/classification/interval_based/_risf.py:73]: should this maybe be repaced with the `capabilities` tags (actual tags), given that this file is new?

File: sktime/registry/_tags.py
Comment (fkiraly) [inline sktime/registry/_tags.py:107]: have you checked whether this is not used in forecasters anymore?

File: sktime/registry/_tags.py
Comment (fkiraly) [inline sktime/registry/_tags.py:228]: this is perhaps still used in transformers that haven't been refactored yet, may cause errors

Comment (fkiraly) [reply sktime/registry/_tags.py:228]: might also be worth adding a todo here

File: sktime/series_as_features/tests/test_all_panel_estimators.py
Comment (fkiraly) [inline sktime/series_as_features/tests/test_all_panel_estimators.py:72]: why are we changing the error message?

File: sktime/classification/base.py
Comment (TonyBagnall) [reply sktime/classification/base.py:150]: its not, left the comment from before, thanks

File: sktime/classification/feature_based/_signature_classifier.py
Comment (TonyBagnall) [reply sktime/classification/feature_based/_signature_classifier.py:105]: good spot, thought I had caught those, wonder if tests will catch it?

File: sktime/classification/interval_based/_risf.py
Comment (TonyBagnall) [reply sktime/classification/interval_based/_risf.py:73]: I'm leaving risf for another day, it does not comply to the interface and is not mine, but will sort it. its a new file but access is the same through init.

File: sktime/registry/_tags.py
Comment (TonyBagnall) [reply sktime/registry/_tags.py:228]: well iI can find no occurrences of coerce-X-to-numpy with find in files in pycharm. Are there any tests anywhere to see if tags declared exist in the registry?

File: sktime/series_as_features/tests/test_all_panel_estimators.py
Comment (TonyBagnall) [reply sktime/series_as_features/tests/test_all_panel_estimators.py:72]: hmmm, not sure! I'll put it back and see if anything breaks

**Review by fkiraly — APPROVED** [review]
Looks good to me now, thanks.

**Comment by fkiraly** [comment]
(we need to wait whether the tests pass)

File: sktime/registry/_tags.py
Comment (TonyBagnall) [reply sktime/registry/_tags.py:107]: yes

**Pull request merged**
